docs/api/en/materials/Material.html

```diff
@@ -417,7 +417,9 @@ <h3>
 		<p>
 			An optional callback that is executed immediately before the shader
 			program is compiled. This function is called with the shader source code
-			as a parameter. Useful for the modification of built-in materials.
+			as a parameter. Useful for the modification of built-in materials, but the recommended 
+      approach moving forward is to use [link:https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language TSL]. 
+      See the before/after example [link:https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language#example here].
```

Comment:
It must be noted that TSL can't be used with `WebGLRenderer`.

Reply:
I didn't know about that! 

The [doc](https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language#why-tsl) says: 

> Other benefits that TSL brings besides simplifying shading creation is keeping the renderer agnostic

And I can't find any instance of "WebGLRenderer" or "WebGPURenderer" on the page.

I'm sure you are right on this as you know Three.js infinitely better than me, so could you elaborate or link to something I can read so that I understand?

It will help me make a clear statement in the doc.

Thanks!

Reply:
We are currently in the process of documenting the node material, TSL and `WebGPURenderer` so if you wait two or three releases there should be more resources in that regard.

We've already done good progress with the node material and TSL but `WebGPURenderer` isn't documented yet. There are also no guides and articles yet which provide a more coherent view on the new system.  But like I said before we have noticed the demand from the developer side and focus on docs now.

To answer the actual question in short: The node material and TSL are only supported by `WebGPURenderer`.  `WebGPURenderer` has a WebGPU and WebGL backend though so it can generate WGSL or GLSL depending on the supported backend. The simplifies the development of 3D apps on the web since you don't need separate renderers and shaders for WebGPU and WebGL.

Reply:
For now, I would rephrase the new text to:

>  Useful for the modification of built-in materials, but the recommended approach moving forward is to use `WebGPURenderer` with the new Node Material system and [link:https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language TSL].

Reply:
> there should be more resources in that regard

Awesome!

> WebGPURenderer has a WebGPU and WebGL backend though so it can generate WGSL or GLSL depending on the supported backend.

There was my confusion! I thought `WebGLRenderer` handled WebGL/GLSL and `WebGPURenderer` was dealing with WebGPU/WGSL. Thanks for clarifying :)

> I would rephrase the new text to:

I implemented you suggestion (en/fr)